compiler/rustc_typeck/src/check/pat.rs

```diff
@@ -772,6 +773,26 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         let msg = "introduce a new binding instead";
                         let sugg = format!("other_{}", ident.as_str().to_lowercase());
                         e.span_suggestion(ident.span, msg, sugg, Applicability::HasPlaceholders);
+
+                        let def_id_opt = match pat_ty.kind {
+                            Adt(def, _) => Some(def.did),
+                            _ => None,
+                        };
+                        let ranges = &[
+                            self.tcx.lang_items().range_struct(),
+                            self.tcx.lang_items().range_from_struct(),
+                            self.tcx.lang_items().range_to_struct(),
+                            self.tcx.lang_items().range_full_struct(),
+                            self.tcx.lang_items().range_inclusive_struct(),
+                            self.tcx.lang_items().range_to_inclusive_struct(),
+                        ];
+                        if ranges.contains(&def_id_opt) {
+                            let msg = 
+                                "Constants only support matching the same type. \
+                                If you are using range syntax (i.e. ..= or similar) in your constant \
+                                you may want to move the range into the match block";
```

Comment:
Maybe some better wording would be:

```suggestion
                                "if you meant to match against a range of values, consider using a range pattern like `min ..= max`";
```

...or something along those lines